### PR TITLE
ingest: processor stats

### LINKED ISSUE #33387: [ingest] Per processor metrics
Currently if trying to review metrics for the ingest node, the lowest granularity is per pipeline [1].  Ideally the metrics would also show per process information too. 

The current behavior of `GET _nodes/stats/ingest`
```json
"nodes": {
  ...
      "ingest": {
        "total": {
          "count": 25,
          "time_in_millis": 8,
          "current": 0,
          "failed": 0
        },
        "pipelines": {
          "mypipeline3": {
            "count": 0,
            "time_in_millis": 0,
            "current": 0,
            "failed": 0
          },
          "mypipeline2": {
            "count": 6,
            "time_in_millis": 0,
            "current": 0,
            "failed": 0
          },
          "mypipeline": {
            "count": 19,
            "time_in_millis": 8,
            "current": 0,
            "failed": 0
          }
        }
      }
    }
```

The proposal is here is to add the same stats as the pipeline has, but also to the processor. For example:

```json
{
   "mypipeline":{
      "count":19,
      "time_in_millis":8,
      "current":0,
      "failed":0,
      "processors":[
         {
            "set":{
            "count":19,
            "time_in_millis":4,
            "current":0,
            "failed":0
            }
         },
         {
            "rename":{
            "count":19,
            "time_in_millis":4,
            "current":0,
            "failed":0
           }
         }
      ]
   }
}
```

~Since each processor can have a on_failure processor, and that on_failure processor can also have multiple processors with other on_failure processors, the resultant JSON can (if pipeline are defined this way) result in quite verbose tree like structures. However, I suspect that most pipeline don't go too deep into setting multiple on_failure handlers.~

EDIT: After further discussion, the on_failure processors should be considered part of the parent processor.

Some recent additions to the ingest node capability should also be addressed here too. (open for discussion)
* conditional `if` https://github.com/elastic/elasticsearch/pull/32398 . Attempts should be made to hide implementation details for the `if` conditional report any metrics that include the time take and/errors produced by the `if` conditional as part of the processor itself. 
* calling other pipelines via the pipeline processor : https://github.com/elastic/elasticsearch/pull/32473 . The pipeline processor should show as any other processor, however, the metrics for `this` pipeline AND `that` pipeline should both increase. 

Also, tag's should be supported in the naming. Likely via the main name, for example `rename:my_tag` if the tag is defined. 

https://www.elastic.co/guide/en/elasticsearch/reference/current/cluster-nodes-stats.html#ingest-stats

### PR DESCRIPTION
**Pull request by jakelandis**
This change introduces stats per processors. Total, time, failed,
current are currently supported. All pipelines will now show all
top level processors that belong to it. Failure processors are not
displayed, however, the time taken to execute the failure chain is part
of the stats for the top level processor.

The processor name is the type of the processor, ordered as defined in
the pipeline. If a tag for the processor is found, then the tag is
appended to the type (colon separated).

Pipeline processors will have the pipeline name appended (colon separated)
to the name of the pipeline (before the tag if one exists). 
If more then one pipeline is used to process the document, then each pipeline
will carry its own stats. The outer most pipeline will also include the
inner most pipeline stats.

Conditional processors will only be included in the stats if the condition
evaluates to true.

Best attempts are made to carry forward processor metrics between 
cluster state changes. If no changes are made to the pipeline, metrics
will be carried forward. However, if the pipeline changes and the
number of processors or the order of the processors (determined by type)
changes, the processor metrics will be reset to zero. 

Closes #33387 

----------

An example pipeline:
```
"test_pipeline": {
  "count": 0,
  "time_in_millis": 0,
  "current": 0,
  "failed": 0,
  "processors": [
    {
      "grok": {
        "count": 0,
        "time_in_millis": 0,
        "current": 0,
        "failed": 0
      }
    },
    {
      "date": {
        "count": 0,
        "time_in_millis": 0,
        "current": 0,
        "failed": 0
      }
    },
    {
      "remove": {
        "count": 0,
        "time_in_millis": 0,
        "current": 0,
        "failed": 0
      }
    }
  ]
}
```
^^ The `  "processors": [` is new. 

An example with a tag and a pipeline processor:
```
"mypipeline1": {
  "count": 0,
  "time_in_millis": 0,
  "current": 0,
  "failed": 0,
  "processors": [
    {
      "set:sets the thing to true": {
        "count": 0,
        "time_in_millis": 0,
        "current": 0,
        "failed": 0
      }
    },
    {
      "pipeline:mypipeline2": {
        "count": 0,
        "time_in_millis": 0,
        "current": 0,
        "failed": 0
      }
    }
  ]
}
```
Rally with the http_logs with grok was run against the code prior and with this change. No performance impact was noticed. 

Also, note this PR does not make any changes to the output of simulate or simulate?verbose. (that will be a different PR).

### CODE REVIEW

**Review by rjernst — COMMENTED** [review]
A few comments

File: server/src/main/java/org/elasticsearch/ingest/ConditionalProcessor.java [hunk 1]
```diff
@@ -42,24 +44,51 @@
...
+                throw e;
+            } finally {
+                long ingestTimeInMillis = clock.millis() - startTimeInMillis;
+                metric.postIngest(ingestTimeInMillis);
+            }
         }
         return ingestDocument;
     }
 
+    Processor getProcessor() {
```
[the remark this answers is not in the document]
Comment (jakelandis) [reply server/src/main/java/org/elasticsearch/ingest/ConditionalProcessor.java]: This is used here: https://github.com/elastic/elasticsearch/pull/34202/files#diff-579dffc1e22e3db13c41f685046b2891R439 to get the actual processor name

File: server/src/main/java/org/elasticsearch/ingest/IngestService.java [hunk 2]
```diff
@@ -262,11 +263,52 @@ public void applyClusterState(final ClusterChangedEvent event) {
                 Pipeline originalPipeline = originalPipelines.get(id);
                 if (originalPipeline != null) {
                     pipeline.getMetrics().add(originalPipeline.getMetrics());
+                    List<Tuple<Processor, IngestMetric>> oldPerProcessMetrics = new ArrayList<>();
+                    List<Tuple<Processor, IngestMetric>> newPerProcessMetrics = new ArrayList<>();
+                    getProcessorMetrics(originalPipeline.getCompoundProcessor(), oldPerProcessMetrics);
+                    getProcessorMetrics(pipeline.getCompoundProcessor(), newPerProcessMetrics);
+                    //Best attempt to populate new processor metrics using a parallel array of the old metrics. This is not ideal since
```
[the remark this answers is not in the document]
Comment (jakelandis) [reply server/src/main/java/org/elasticsearch/ingest/IngestService.java]: When 1 pipeline changes, then all pipelines are rebuilt and without this code we would loose all metrics on any pipeline change. For example we have 2 pipelines, and 1 one of them is deleted, we don't want to loose the metrics for the pipeline that had no changes.  We don't have easy access to exactly which pipeline changed, and the heuristic to carry forward the metrics is if the pipeline still exists, the count of processors and the types of processor (in order) don't change, then carry forward the metrics.

File: server/src/main/java/org/elasticsearch/ingest/CompoundProcessor.java [hunk 3]
```diff
@@ -94,12 +114,17 @@ public String getTag() {
 
     @Override
     public IngestDocument execute(IngestDocument ingestDocument) throws Exception {
-        for (Processor processor : processors) {
+        for (Tuple<Processor, IngestMetric> processorWithMetric : processorsWithMetrics) {
+            Processor processor = processorWithMetric.v1();
+            IngestMetric metric = processorWithMetric.v2();
+            long startTimeInMillis = clock.millis();
```
[the remark this answers is not in the document]
Comment (jakelandis) [reply server/src/main/java/org/elasticsearch/ingest/CompoundProcessor.java]: @rjernst I removed the `clock` in favor a LongSupplier backed by System:nanoTime https://github.com/elastic/elasticsearch/pull/34202/commits/27143386b54a4d5775b0dc8e5e7b680981dc2a40. Can you please confirm that this is the correct fix here ? I made the same mistake in a different PR I will need fix too.

File: server/src/main/java/org/elasticsearch/ingest/IngestStats.java [hunk 4]
```diff
@@ -71,22 +97,43 @@ public Stats getTotalStats() {
     }
 
     /**
-     * @return The stats on a per pipeline basis
+     * @return The stats on a per pipeline basis.  A Map(pipelineId -> Tuple(pipelineStats, List(perProcessorStats))
+     * where perProcessorStats = Tuple(processorDisplayName, processorStats)
      */
-    public Map<String, Stats> getStatsPerPipeline() {
+    public Map<String, Tuple<IngestStats.Stats, List<Tuple<String, IngestStats.Stats>>>> getStatsPerPipeline() {
```
[the remark this answers is not in the document]
Comment (jakelandis) [reply server/src/main/java/org/elasticsearch/ingest/IngestStats.java]: Fixed in https://github.com/elastic/elasticsearch/pull/34202/commits/8067758d4803b892381f24842ca1b64b7761f899. IngestStats now accepts totalStats, List<PipelineStat>, and Map<String, ProcessorStat> (keyed by pipelineId), and a builder to help build from the Metrics representation. I hope this makes the code more readable.

**Comment by jakelandis** [comment]
@rjernst - All initial comments have been addressed. Mind to take another look ?

File: server/src/main/java/org/elasticsearch/ingest/CompoundProcessor.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (jakelandis) [reply server/src/main/java/org/elasticsearch/ingest/CompoundProcessor.java]: @rjernst - I also fixed this for other places where I made the same mistake: https://github.com/elastic/elasticsearch/pull/34202/commits/ff98a82d158d71ab4bd7ffc367c0e627f1d6d453

**Review by rjernst — APPROVED** [review]
LGTM

**Comment by jasontedor** [comment]
I reverted this from master in 0577703183ec3e341b88ff3ff43410683f5bbad9 due to failing tests in the mixed cluster tests.

**Comment by jakelandis** [comment]
Re-introduce this change with test fix on #34724

**Pull request merged**